core/lib/lantern/page-dependency-graph.js

```diff
@@ -439,6 +535,258 @@ class PageDependencyGraph {
+   */
+  static chooseInitiatorRequest(record, recordsByURL) {
+    if (record.redirectSource) {
+      return record.redirectSource;
+    }
+
+    const initiatorURL = PageDependencyGraph.getNetworkInitiators(record)[0];
+    let candidates = recordsByURL.get(initiatorURL) || [];
+    // The (valid) initiator must come before the initiated request.
+    candidates = candidates.filter(c => {
+      return c.responseHeadersEndTime <= record.rendererStartTime &&
+          c.finished && !c.failed;
+    });
+    if (candidates.length > 1) {
+      // Disambiguate based on prefetch. Prefetch requests have type 'Other' and cannot
+      // initiate requests, so we drop them here.
+      const nonPrefetchCandidates = candidates.filter(
+          cand => cand.resourceType !== RESOURCE_TYPES.Other);
+      if (nonPrefetchCandidates.length) {
+        candidates = nonPrefetchCandidates;
+      }
+    }
+    if (candidates.length > 1) {
+      // Disambiguate based on frame. It's likely that the initiator comes from the same frame.
+      const sameFrameCandidates = candidates.filter(cand => cand.frameId === record.frameId);
+      if (sameFrameCandidates.length) {
+        candidates = sameFrameCandidates;
+      }
+    }
+    if (candidates.length > 1 && record.initiator.type === 'parser') {
+      // Filter to just Documents when initiator type is parser.
+      const documentCandidates = candidates.filter(cand =>
+        cand.resourceType === RESOURCE_TYPES.Document);
+      if (documentCandidates.length) {
+        candidates = documentCandidates;
+      }
+    }
+    if (candidates.length > 1) {
+      // If all real loads came from successful preloads (url preloaded and
+      // loads came from the cache), filter to link rel=preload request(s).
+      const linkPreloadCandidates = candidates.filter(c => c.isLinkPreload);
+      if (linkPreloadCandidates.length) {
+        const nonPreloadCandidates = candidates.filter(c => !c.isLinkPreload);
+        const allPreloaded = nonPreloadCandidates.every(c => c.fromDiskCache || c.fromMemoryCache);
+        if (nonPreloadCandidates.length && allPreloaded) {
+          candidates = linkPreloadCandidates;
+        }
+      }
+    }
+
+    // Only return an initiator if the result is unambiguous.
+    return candidates.length === 1 ? candidates[0] : null;
+  }
+
+  /**
+   * @param {LH.TraceEvent[]} mainThreadEvents
+   * @param {LH.Trace} trace
+   * @param {LH.Artifacts.TraceEngineResult} traceEngineResult
+   * @param {LH.Artifacts.URL} URL
+   */
+  static async createGraphFromTrace(mainThreadEvents, trace, traceEngineResult, URL) {
+    if (trace._testSmuggledNetworkRecords) {
+      const lanternRequests = trace._testSmuggledNetworkRecords;
+      const graph = PageDependencyGraph.createGraph(mainThreadEvents, lanternRequests, URL);
+      return {graph, records: lanternRequests};
+    }
+
+    // TODO: trace engine should handle this.
+    const serviceWorkerThreads = new Map();
```

Comment:
extract this and from `fromWorker` bit into a ~`setFromWorker` fn. i'm fine if this fn remains inside of createGraphFromTrace..